### DISCUSSION TITLE
SMARTS matching problem

### DISCUSSION THREAD

**Discussion by victorsabanza** [rdkit/rdkit#7453] [Q&A] [answered]
Hi, I have a problem concerning SMARTS pattern finding.

I have the following code

```
mol1 = Chem.MolFromSmiles('Oc1cc(C(O)C(Oc2ccccc2(OC))CO)cc(OC)c1OC')
mol2 = Chem.MolFromSmiles('COc1cc(C(O)C(Oc2c(OC)cccc2(OC))CO)cc(OC)c1OC')

patt1 = Chem.MolFromSmarts('COc1c(OC)c(O)cc(CCOc2c(OC)cccc2)c1')

ats = [mol.GetSubstructMatch(patt1) for mol in [mol1, mol2]]

Chem.Draw.MolsToGridImage(mols=[mol1, mol2], highlightAtomLists=ats)
```
Matching the two molecules.
![image](https://github.com/rdkit/rdkit/assets/95976192/09e044a6-e2ca-47a7-a76c-afc49d0a1b95)

I would like to edit the SMARTS in patt1 so it only matches the first molecule based on the "ortho" position of the first phenyl ring (the one on the left), which could bear anything but an oxygen atom. I tried these and several modifications without success (no match for any of the molecules):

```
# match based on negating oxygen presence
patt2 = Chem.MolFromSmarts('COc1c(OC)c(O)cc(CCOc2c(OC)cccc2[!#8])c1')

# match based on only C, c or H presence
patt3 = Chem.MolFromSmarts('COc1c(OC)c(O)cc(CCOc2c(OC)cccc2[C,c,#1])c1')

```
<img width="682" alt="Screenshot 2024-05-21 at 21 12 38" src="https://github.com/rdkit/rdkit/assets/95976192/dc145448-8324-411f-9303-e8b7ab2f671d">

Is there something tricky I am missing? Thanks in advance!

**Comment by greglandrum** [comment, marked as answer]
Hi @victorsabanza 
This is an ideal situation to use a recursive SMARTS: you want to say that the atom is an aromatic carbon, but not an aromatic carbon connected to an O. That looks like this: `COc1c(OC)c(O)cc(CCOc2c(OC)ccc[c;!$(c-O)]2)c1`

Here it is for your two molecules:
```
In [13]: patt2 = Chem.MolFromSmarts('COc1c(OC)c(O)cc(CCOc2c(OC)ccc[c;!$(c-O)]2)c1')

In [14]: print(mol1.HasSubstructMatch(patt2),mol2.HasSubstructMatch(patt2))
True False
```

**Reply by victorsabanza** [reply]
It works perfectly, thanks a lot for the suggestion! @greglandrum
